### PR TITLE
Add fmtlib for log formatting

### PR DESCRIPTION
**Pull request by fsimonis**
## Main changes of this PR

Add [fmtlib](fmt.dev) to simplify formatting of error messages.
Changed log macros to format all arguments directly to `fmt::format()`.
Ported all log calls to the new syntax.

```cpp
PRECICE_XXX(message, [arg0], [arg1])
PRECICE_CHECK(cond, message, [arg0], [arg1])

PRECICE_INFO("Hello");
PRECICE_DEBUG("The polynomial linear system of the RBF mapping from mesh {} to mesh {} converged. {}",
              input()->getName(), output()->getName(), _QRsolver.summaryFor(tau));

PRECICE_CHECK(fromMesh.get() != nullptr,
              "Mesh \"{0}\" was not found while creating a mapping. "
              "Please correct the from=\"{0}\" attribute.",
              fromMeshName);
```

Short term goals:
* Easier to format error messages in code.
* More readable error messages in code.
* Prevent duplicate arguments in the formatting using positional arguments `{0} x {0}`

Long term goal:
* Drop `prettyprint` dependency.  
  (Still required by assertions and some additional places)
* Simplify formatting in exporters
* Replace by `std::format` in the future

## Motivation and additional information

* Formatting error messages is currently a tedious process as messages have to be interleaved with `<<`.
* This results in messages that are difficult to format and difficult to read.
* `clang-format` attempts to align `<<` which makes formatting even more tedious. 

## Author's checklist

* [ ] I added a changelog file with this PR number in `docs/changelog/` if there are noteworthy changes.
* [ ] I ran `tools/formatting/check-format` and everything is formatted correctly.
* [x] I sticked to C++14 features.
* [x] I sticked to CMake version 3.10.
* [ ] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [x] Do you understand the code changes?
* [ ] (more questions/tasks)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -188,6 +188,8 @@ else()
 endif()
 precice_validate_prettyprint()
 
+# libfmt
+add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/thirdparty/fmt)
```
Comment (DavidSCN) [inline CMakeLists.txt:192]: That's only for the libraray. You need to add an include directory for the binary and tests as well in order to compile
```cmake 
  include_directories(${CMAKE_CURRENT_SOURCE_DIR}/thirdparty/fmt/include)
```

**Review by DavidSCN — COMMENTED** [review]
Very nice, this improves the readability of the error messages in the code and writing error messages itself by a lot. Still, my long-term goal is to introduce a column limit at some time. 
Just one more comment: wouldn't it be better to use a git submodule or something similar instead of a hard copy? The `fmt` code is frequently updated and this way we could simply pull the latest changes into our code base.

File: CMakeLists.txt (hunk 1 above)
Comment (fsimonis) [reply CMakeLists.txt:192]: This processes the [`CMakeLists.txt`](https://github.com/precice/precice/blob/b4aa52aca2c086782d02badc1c60955ba5593cf4/thirdparty/fmt/CMakeLists.txt) in the `thirdparty/fmt` directory.

**Comment by fsimonis** [comment]
@DavidSCN The boost logger should take care of the linebreak at the end of log statements.
Do you have an example for such an issue?

File: CMakeLists.txt (hunk 1 above)
Comment (DavidSCN) [reply CMakeLists.txt:192]: Still, it didn't compile on system without adding `include_directories`. Have you enabled testing and the binary building?

**Comment by DavidSCN** [comment]
There is a line break, but I was thinking of an additional empty line. It changed now due to the restructuring, but an OpenFOAM error looked from the user perspective like this:
```
---[precice] ERROR:  The given mesh name "Fluid-Meshs" is unknown to preCICE. Please check the mesh definitions in the configuration.
Looking for any time directories without results (e.g. stray functionObjectProperties files, see openfoam-adapter issue #26 on GitHub)...
```
so there is no clear separation. But thinking about it again, maybe we should keep it like this. (1) The restructuring prevents such occurrences and (2) an empty line might be confusing if nothing follows.

**Comment by fsimonis** [comment]
> But thinking about it again, maybe we should keep it like this. (1) The restructuring prevents such occurrences and (2) an empty line might be confusing if nothing follows.

I agree, however you can experiment with adding `\n` in the  [message format](https://www.precice.org/configuration-logging.html#examples).

File: CMakeLists.txt (hunk 1 above)
Comment (fsimonis) [reply CMakeLists.txt:192]: I tested it what is the error message?

The include directory is propagated by the target.

Comment (DavidSCN) [reply CMakeLists.txt:192]: Tested it now as well in a fresh build and it works for me as well. I previously rebuild an old build, but still resourced the root CMakeLists.txt in order to incorporate the new include directory. I have no idea why. Sorry for the mess.

**Review by MakisH — APPROVED** [review]
- I like this proposal: it can make complex formatting easier, but mostly in the long term.
- Header-only, so it should not change anything for the user.
- I understand the changes, but I have not checked every file.
- The license (see `thirdparty/fmt/include/fmt/format.h` or `thirdparty/fmt/LICENSE.rst`) seems to be permissive, I think BSD, so it should be fine with our LGPL).
- I have not tried it, I woud just trust the CI.

TODO: Add a changelog entry.

Questions:
- Is the library C++14? We need to compile it with the same compiler.
- Is there anything to consider regarding xSDK and symbols we export when using header-only libraries?

Let me know if you would like me to look deeper into anything.

**Comment by fsimonis** [comment]
@MakisH 
> Is the library C++14? We need to compile it with the same compiler.

It is actually C++11. So we could have started to use this a long time ago :wink: 

> Is there anything to consider regarding xSDK and symbols we export when using header-only libraries?

Not AFAI, however, we already use Eigen in a header-only way. So there won't be new The solution to this is #623 btw.

**Pull request merged**